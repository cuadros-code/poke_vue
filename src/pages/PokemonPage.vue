<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>Quién es este pokémon?</h1>
    <PokemonImage :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions
      :options="pokemonArr" 
      @selection="checkAnswer"
    />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo juego
      </button>
    </template>
  </div>
</template>

<script>
import PokemonImage from '@/components/PokemonImage.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import getPokemonsOptions from '@/helpers/getPokemonOptions'

export default {
  name: 'PokemonPage',
  components: {
    PokemonImage,
    PokemonOptions
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  mounted() {
    this.mixPokemonArray()
  },
  methods: {
    async mixPokemonArray(){
      this.pokemonArr = await getPokemonsOptions()
      const rndInt = Math.floor( Math.random() * 4 )
      this.pokemon = this.pokemonArr[rndInt]
    },
    checkAnswer( pokemonId ) {
      this.showPokemon = true
      this.showAnswer = true
      pokemonId === this.pokemon.id 
        ? this.message = `Correcto, ${this.pokemon.name}`
        : this.message = `Oops, era ${this.pokemon.name}`
    },
    newGame() {
      this.mixPokemonArray()
      this.showPokemon = false
      this.showAnswer = false
      this.pokemon = null
    }
  },
}
</script>
<style scoped>
h2{
  padding: 0;
  margin: 0 0 10px 0;
}
</style>