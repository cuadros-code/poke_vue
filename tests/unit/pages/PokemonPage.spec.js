import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'

describe('PokemonPage', () => { 

  let wrapper

  beforeEach(() => {
    wrapper = shallowMount( PokemonPage )
  })

  test('Debe de llamar el mixPokemon', () => {

    const mixPokemonArr = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
    shallowMount( PokemonPage )
    expect( mixPokemonArr ).toHaveBeenCalled()
    
  })

  test('Debe de hacer macth al cargar los pokemons', () => { 

    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr: [
            { name: 'bulbasaur', id: 1 },
            { name: 'ivysaur', id: 2 },
            { name: 'venusaur', id: 3 },
            { name: 'charmander', id: 4 }
          ],
          pokemon: { name: 'bulbasaur', id: 1 },
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    expect( wrapper.html() ).toMatchSnapshot()

  })

  test('Debe de mostrar los componentes pokemonImage pokemonOptions', () => { 

    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr: [
            { name: 'bulbasaur', id: 1 },
            { name: 'ivysaur', id: 2 },
            { name: 'venusaur', id: 3 },
            { name: 'charmander', id: 4 }
          ],
          pokemon: { name: 'bulbasaur', id: 1 },
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    expect( wrapper.getComponent('pokemon-image-stub' ) ).toBeTruthy()
    expect( wrapper.getComponent('pokemon-options-stub' ) ).toBeTruthy()

    expect( wrapper.getComponent('pokemon-image-stub' ).attributes('pokemonid') ).toBe("1")

    expect( wrapper.getComponent('pokemon-options-stub' ).attributes('options') ).toBeTruthy()

  })


  test('Pruebas checkAnswer', async () => { 
    
    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr: [
            { name: 'bulbasaur', id: 1 },
            { name: 'ivysaur', id: 2 },
            { name: 'venusaur', id: 3 },
            { name: 'charmander', id: 4 }
          ],
          pokemon: { name: 'bulbasaur', id: 1 },
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    await wrapper.vm.checkAnswer(2)

    expect( wrapper.find('h2').exists() ).toBeTruthy()

    expect( wrapper.vm.showPokemon ).toBeTruthy()

  })

})