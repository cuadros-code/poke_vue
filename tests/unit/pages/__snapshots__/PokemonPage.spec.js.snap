// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Debe de hacer macth al cargar los pokemons 1`] = `
<div>
  <h1>Quién es este pokémon?</h1>
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub options="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
